### ISSUE TITLE
Inductor excessively materializes views in backward

### ISSUE THREAD

**Issue by ngimel** [pytorch/torchdynamo#595]
Repro in https://gist.github.com/ngimel/887241bb5d7bb34adae4f9a329dcfea6 (includes python source, AOT graphs and compiled program), originally encountered in hf_Bert model. Repro doesn't generate exact same code as full model generates, but is enough to reproduce a problem. 
Repro is nn.Linear layer followed by an activation (GeLU or ReLU, ReLU generates simpler code but still has this problem). This is an excessively common pattern. In eager with ReLU activation, non-gemm kernels take approx 200 us on A100, inductor generated non-gemm kernels are 400 us, in large part due to excessive view materialization. 
In backward, gradInput for the activation is used to compute gradients of bias, weight and input. AOT graph shows that for weight and input gradient computation, a view or permute op is performed that doesn't involve data movement. Inductor generates code that copies activation gradInput (twice) instead of taking a view of it. 
I don't quite know how to fix it, so advice welcome.

**Comment by ngimel** [comment]
Here's another one where too many buffers are realized in forward. It's linear - gelu - linear. https://gist.github.com/ngimel/1a7156a98a5bebab31e15f5b5f6b222d
`buf1` which is the result of mm+bias is an output (even though it should be recomputed in backward
`buf2` and `buf5` contain the same output, `buf2` is used in the next mm, `buf5` is an output.
